Translation FAQ: Passing Functions to Components (#108)
reactjs/ru.reactjs.org: content/docs/faq-functions.md

}
```

> #### Привязка в свойствах класса (Этап 3: кандидаты) {#class-properties-stage-3-proposal}

Reviewer: Проще:

```suggestion
#### Привязка в свойствах класса (предложение-кандидат) {#class-properties-stage-3-proposal}
```

}
```

>**Примечание:**
>

> >Использование `Function.prototype.bind` в render() создаёт новую функцию при каждой отрисовке компонента, что может повлиять на производительность (см. ниже).

Reviewer: Вместо _отрисовки_ и производным ему используем _рендер_, например:

```suggestion
>Использование `Function.prototype.bind` в render() создаёт новую функцию при каждом рендере компонента, что может повлиять на производительность (см. ниже).
```

Можете, пожалуйста, исправить это по всему тексту?

}
```

>**Примечание:**
>
>Использование стрелочной функции в render() создает новую функцию при каждой отрисовке компонента, что может повлиять на производительность (см. ниже).

### Можно ли использовать стрелочные функции в методе render()? {#is-it-ok-to-use-arrow-functions-in-render-methods}

> В целом, да. Зачастую это самый простой способ передать параметры через функции-колбэки.

Reviewer: Проще:

```suggestion
В целом, да. Зачастую это самый простой способ передать параметры через колбэки.
```

method();
```

Привязка гарантирует, что второй фрагмент будет работать так же, как и первый. 

В React, как правило, привязывать нужно только те методы, которые вы *хотите передать* другим компонентам. Например, `<button onClick={this.handleClick}>` передаёт `this.handleClick`, поэтому его нужно привязать. Впрочем, метод `render` и методы жизненного цикла привязывать не обязательно, так как мы не передаём их в другие компоненты.

> [Ознакомьтесь со статьёй Йехуды Катц](http://yehudakatz.com/2011/08/11/understanding-javascript-function-invocation-and-this/), которая более подробно объясняет, что такое привязка, и как работают функции в JavaScript.

Reviewer 2: ```suggestion
[Ознакомьтесь со статьёй Йехуды Катц](http://yehudakatz.com/2011/08/11/understanding-javascript-function-invocation-and-this/), в которой более подробно объяснено, что такое привязка, и как работают функции в JavaScript.
```

Reviewer 2: Я сперва понял _"которая объясняет"_ как если бы было написано что-то вроде _Йехуда Катц которая объясняет_, хотя Йехуда, как я понимаю, не женщина. 😅

### Почему моя функция вызывается каждый раз при отрисовке компонента? {#why-is-my-function-being-called-every-time-the-component-renders}

Убедитесь, что вы не _вызываете функцию_, когда передаёте ее компоненту:

```jsx
render() {

>   // Неправильно: вместо ссылки, функция handleClick была вызвана!

Reviewer: ```suggestion
  // Неправильно: вместо ссылки была вызвана функция handleClick!
```

  return <button onClick={this.handleClick()}>Click Me</button>
}
```

> Вместо этого, *просто передайте функцию* (без скобок):

Reviewer: ```suggestion
Вместо этого *просто передайте функцию* (без скобок):
```

> Вместо этого, *просто передайте функцию* (без скобок):

Reviewer 2: Есть мнение, что "просто" -- слово-паразит. В паре моих PR они были вычищены.

Author: Вычистим!

Reviewer 2: Можно сказать

> Вместо этого передайте *саму функцию* (без скобок):

```jsx
render() {
  // Правильно: handleClick передаётся как ссылка!
  return <button onClick={this.handleClick}>Click Me</button>
}
```

### Как передать параметры обработчику событий или колбэку? {#how-do-i-pass-a-parameter-to-an-event-handler-or-callback}

Чтобы передать параметры обработчику событий, оберните его в стрелочную функцию:

```jsx
<button onClick={() => this.handleClick(id)} />
```

> Это действие равно использованию `.bind`:

Reviewer 2: ```suggestion
Это действие равносильно использованию `.bind`:
```

```jsx
<button onClick={this.handleClick.bind(this, id)} />
```

> #### Пример: Передача параметров с использованием стрелочных функций{#example-passing-params-using-arrow-functions}

Reviewer: ```suggestion
#### Пример: Передача параметров с использованием стрелочных функций {#example-passing-params-using-arrow-functions}
```

```jsx

> const A = 65 // ASCII код символа

Reviewer: Упустил из виду :man_facepalming: 

```suggestion
const A = 65 // ASCII-код символа
```

}
```

#### Пример: Передача параметров с использованием атрибутов данных {#example-passing-params-using-data-attributes}

В качестве альтернативного подхода вы можете использовать DOM API, чтобы хранить необходимые для обработчиков событий данные. Рассмотрите этот подход, если вам нужно оптимизировать большое количество элементов или использовать дерево визуализации, полагающееся на компонент React.PureComponent для проверки на равенство.

```jsx

> const A = 65 // ASCII код символа

Reviewer: ```suggestion
const A = 65 // ASCII-код символа
```

}
```

### Как предотвратить слишком быстрый или слишком частый вызов функции? {#how-can-i-prevent-a-function-from-being-called-too-quickly-or-too-many-times-in-a-row}

Если вы используете обработчики событий, такие как `onClick`  или `onScroll`, и хотите предотвратить быстрое срабатывание колбэков, вы можете ограничить скорость выполнения колбэка. Для этого вы можете использовать: 

> - **throttling**: выборочные изменения, зависимые от частоты, основанной на времени (напр. [`_.throttle`](https://lodash.com/docs#throttle))

Reviewer 2: Думаю, тут можно перевести throttling и debouncing тоже.

- **дебаунсинг**: изменения, задействованные после некого периода бездействия (напр. [`_.debounce`](https://lodash.com/docs#debounce))

> - **Тротлинг с помощью `requestAnimationFrame`**: выборочные изменения, основанные на [`requestAnimationFrame`](https://developer.mozilla.org/ru/docs/DOM/window.requestAnimationFrame) (напр. [`raf-schd`](https://github.com/alexreardon/raf-schd))

Reviewer 2: ```suggestion
- **тротлинг с помощью `requestAnimationFrame`**: выборочные изменения, основанные на [`requestAnimationFrame`](https://developer.mozilla.org/ru/docs/DOM/window.requestAnimationFrame) (напр. [`raf-schd`](https://github.com/alexreardon/raf-schd))
```

> Взгляните на [данную визуализацию](http://demo.nimius.net/debounce_throttle/), где сравниваются функции `throttle` и `debounce`.

Reviewer 2: > Взгляните на [визуализацию](http://demo.nimius.net/debounce_throttle/), которая сравнивает функции `throttle` и `debounce`.

> Примечание:
>
> `_.debounce`, `_.throttle` и `raf-schd` предусматривают метод `cancel` для отмены отложенных колбэков. Вы должны либо вызвать этот метод из componentWillUnmount, _либо_ удостоверится, что компонент все еще встроен в пределах отложенной функции.

#### Throttle {#throttle}

> Тротлинг предотвращает повторный вызов функции в заданный период времени. Этот метод был задействован в примере ниже, чтобы не допустить вызов обработчика "click" чаще чем раз в секунду.

Reviewer: Обычно используется вариант с двумя "т" (_Троттлинг_)

Author: @lex111, в предложенном словаре веб-терминов одна 'т': https://github.com/web-standards-ru/dictionary/blob/master/dictionary.md#throttling
Будем с двумя писать?

Reviewer: Упс, не посмотрел, тогда следуем _новому_ стилю, оставляем без изменений :)

Если вы используете обработчики событий, такие как `onClick`  или `onScroll`, и хотите предотвратить быстрое срабатывание колбэков, вы можете ограничить скорость выполнения колбэка. Для этого вы можете использовать: 

- **тротлинг**: выборочные изменения, зависимые от частоты, основанной на времени (напр. [`_.throttle`](https://lodash.com/docs#throttle))
- **дебаунсинг**: изменения, задействованные после некого периода бездействия (напр. [`_.debounce`](https://lodash.com/docs#debounce))

> - **`requestAnimationFrame` throttling**: выборочные изменения, основанные на [`requestAnimationFrame`](https://developer.mozilla.org/ru/docs/DOM/window.requestAnimationFrame) (напр. [`raf-schd`](https://github.com/alexreardon/raf-schd))

Reviewer 2: тут тоже throttling остался на английском.

> Троттлинг с помощью `requestAnimationFrame`

  }

  handleChange(e) {
    // React помещает события в пул, поэтому значение считывается перед debounce.

>     // Как альтернативу, мы могли бы вызвать `event.persist()` и передать событие целиком.

Reviewer: ```suggestion
    // В качестве альтернативы мы могли бы вызвать `event.persist()` и передать событие целиком.
```

}
```

#### Тестирование ограничения скорости {#testing-your-rate-limiting}

> Когда тестирование вашего кода ограничения скорости работает правильно, было бы полезно иметь возможность прокрутить время. Если вы используете [`jest`](https://facebook.github.io/jest/), то для этого вам может пригодиться [`mock timers`](https://facebook.github.io/jest/docs/en/timer-mocks.html). Если вы используете `requestAnimationFrame`, то [`raf-stub`](https://github.com/alexreardon/raf-stub) может оказаться полезным инструментом для управления смены кадров анимации.

Reviewer 2: ```suggestion
Когда вы тестируете, что ваш код ограничения скорости работает правильно, полезно иметь возможность прокрутить время. Если вы используете [`jest`](https://facebook.github.io/jest/), вам может пригодиться [`mock timers`](https://facebook.github.io/jest/docs/en/timer-mocks.html). Если вы используете `requestAnimationFrame`, то [`raf-stub`](https://github.com/alexreardon/raf-stub) может оказаться полезным инструментом для управления смены кадров анимации.
```
